### PR TITLE
Credential xfer

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: bedrock/load-data.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+var testBaseUri = 'https://example.com/credentials/';
+function createUniqueCredential() {
+  var newCredential = bedrock.tools.clone(credentialTemplate);
+  newCredential.id = testBaseUri + bedrock.tools.uuid();
+  return newCredential;
+};
+
+function insertTestData(numberOfRecords, callback) {
+  async.times(numberOfRecords, function(n, next) {
+    store.provider.insert(null, createUniqueCredential(),
```
Comment (dlongley) [inline bedrock/load-data.js]: Readability/style nits: If you have to drop at least one parameter to the next line in order to break before 80 chars, drop all of them (move the first two parameters above down a line and indent them over).

File: bedrock/load-data.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  },
+  signature : {
+    type : 'GraphSignature2012',
+    created : '2015-07-24T12:48:38Z',
+    creator : 'https://example.com/keys/1',
+    signatureValue : 'lRBljDguLA316oTkXoHPxSFYziXTvSZn1Ap2IEZkDc0F93V5BN' +
+      'jHXtC+YS7SbwnYfgBb2d4WnvXDSxzGboAEEw/Jcc2/rz0uqfU1/Jbwps5pLMWnHS/' +
+      '5JY+9PPbHNS8PZSeonpEH2hTvK+ofv6CVu7voF3PK3q/Jw3tjmJ88XTA='
+  }
+};
```
Comment (dlongley) [inline bedrock/load-data.js:41]: Nits: There's an extra space before each colon in the JSON above.

File: bedrock/test.js [hunk 3]
```diff
@@ -2,8 +2,26 @@
...
 var bedrock = require('bedrock');
+var config = bedrock.config;
 
 require('bedrock-express');
 require('../lib/issuer');
 
+// Loads a fake credential into the database
+require('./load-data');
+
+bedrock.events.on('bedrock.configure', function() {
```
Comment (dlongley) [inline bedrock/test.js:16]: Why is this event being used instead of setting the config directly?

File: components/main.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+define([
+  'angular',
+  'underscore',
+  'did-io'
+], function(
+  angular, _, didio
+) {
```
Comment (dlongley) [inline components/main.js]: Nits: Seems like the above could all be put on one line.

File: lib/issuer.js [hunk 5]
```diff
@@ -7,13 +7,25 @@
...
 var brPermission = require('bedrock-permission');
 var database = require('bedrock-mongodb');
+var httpSignature = require('http-signature');
+var identity = require('bedrock-identity');
 var jsigs = require('jsonld-signatures')({
   inject: {jsonld: bedrock.jsonld}
 });
+var ensureAuthenticated = require('bedrock-passport').ensureAuthenticated;
+var scheduler = require('bedrock-jobs');
+// FIXME: bedrock-views is being included to correct express error handling
```
Comment (dlongley) [inline lib/issuer.js]: This should no longer be the case; can we remove bedrock-views now?

File: lib/issuer.js [hunk 6]
```diff
@@ -28,6 +40,36 @@ module.exports = api;
...
+    callback(err);
+  });
+});
+
+bedrock.events.on('bedrock.test.configure', function() {
+  // load test config
+  require('./test.config');
+});
+
+/* Not needed
```
Comment (dlongley) [inline lib/issuer.js]: Can this be removed now?

File: lib/issuer.js [hunk 7]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
-        fields: {recipient: 1, 'credential.sysState': 1, id: 1},
-        options: {unique: true, background: false}
-      }, {
-        collection: store.provider.name,
-        fields: {issuer: 1, 'credential.sysState': 1, id: 1},
-        options: {unique: true, background: false}
-      }], callback);
+// TODO: more acceptable formats?
+function checkHeaders(req, res, next) {
+  var acceptableContentTypes = [
```
Comment (dlongley) [inline lib/issuer.js:228]: Use res.format. eg: https://github.com/digitalbazaar/bedrock-idp/blob/11dd3b0bd27b52edaa1e15b4d3cb73ebe4ece8b9/lib/services.identity.js#L343-L349

File: lib/issuer.js [hunk 8]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+    'application/json',
+    'application/ld+json',
+    'text/html'
+  ];
+  if(!req.accepts(acceptableContentTypes)) {
+    res.status(406);
+    res.write(JSON.stringify({
```
Comment (dlongley) [inline lib/issuer.js]: I think `res.json()` can just take a JS object directly (no need to make the extra stringify call).

File: lib/issuer.js [hunk 9]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+      message: 'Requested content types not acceptable.',
+      details: {
+        accepted: req.get('accept'),
+        acceptable: acceptableContentTypes
+      }
+    }));
+    res.end();
```
Comment (dlongley) [inline lib/issuer.js]: No need to call `res.end` if you use `res.send`/`res.json`.

File: lib/issuer.js [hunk 10]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+    return next();
```
Comment (dlongley) [inline lib/issuer.js]: Don't call the next handler if the request has been handled.

File: lib/issuer.js [hunk 11]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+  }
+  if(acceptableContentTypes.indexOf(req.get('Content-Type')) == -1) {
+    res.status(415).end();
```
Comment (dlongley) [inline lib/issuer.js]: Replace line with `res.sendStatus(415);`

File: lib/issuer.js [hunk 12]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+      message: 'Requested content types not acceptable.',
+      details: {
+        accepted: req.get('accept'),
+        acceptable: acceptableContentTypes
+      }
+    }));
+    res.end();
+    return next();
+  }
+  if(acceptableContentTypes.indexOf(req.get('Content-Type')) == -1) {
```
Comment (dlongley) [inline lib/issuer.js]: Use triple equals.

Comment (dlongley) [inline lib/issuer.js]: Why is request 'Content-Type' (this is for the mime type of the request body) being compared to acceptable content types? Should this be using the `Accept` header instead? How is this check different (or meant to be different) from the `req.accepts` check above? Are the accept type and content type being bound together somehow here? (confusion).

File: lib/issuer.js [hunk 13]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+    res.status(415).end();
+    return next();
+  }
+  next();
+};
+
+function findCredentialByClaimId(claimId, callback) {
+  console.log('CLAIMID:', claimId);
+  query = {
+    'credential.claim.id': claimId
```
Comment (dlongley) [inline lib/issuer.js:253]: Shouldn't this be done using the `recipient` index? If not, why not?

File: lib/issuer.js [hunk 14]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+        public: true
+      }));
     }
-  }, function(err) {
-    callback(err);
+    return next();
+  }
+
+  // FIXME: determine proper endpoint
+  app.get('/creds/claimid/:claimid', function(req, res, next) {
```
Comment (dlongley) [inline lib/issuer.js:277]: Does `bedrock-credentials-rest` provide these end points? cc: @davidlehn

File: lib/issuer.js [hunk 15]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+            'public': true
+          }, err));
+      }
+      return res.json(results);
+    });
+  });
+  // FIXME: this is a provisional endpoint and authentication is required
+  app.put('/creds/:id', function(req, res, next) {
+    var credentialId = req.params.id;
+    console.log('BODY *****', credentialId, req.body);
```
Comment (dlongley) [inline lib/issuer.js]: Remove console.log usage. Though, if `bedrock-credentials-rest` provides this end point already, just remove it and use that. cc: @davidlehn

File: lib/issuer.js [hunk 16]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+            return next(new BedrockError(
+              'The credential could not be updated.',
+              'UpdateCredentialFailed', {
+                httpStatusCode: 400,
+                'public': true
+              }, err));
+          }
+          return res.json({status: 'OK'});
+        });
+    } else {
```
Comment (dlongley) [inline lib/issuer.js]: Style/Readbility nits: If you've got a simple else block, invert the if conditional, move it to the top and return early, then you can unindent the larger if block. Readers can see weed out the simpler logic paths more quickly and then get to the more complicated ones last (and there will be less indenting).

File: lib/issuer.js [hunk 17]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
-});
+  app.post(config.issuer.endpoints.unsignedCredential,
+    [checkHeaders, validate({body:'credential'}), ensureAuthenticated],
+    function(req, res, next) {
+      // FIXME: Replace credential signing key from config with key from db
+      var options = {};
+      options.privateKey = config.issuer.credentialSigningPrivateKey;
+      options.publicKeyId = config.issuer.credentialSigningPublicKeyId;
+      var credential = req.body;
+      api.issue(null, credential, options, function(err) {
```
Comment (dlongley) [inline lib/issuer.js:346]: Add a FIXME that indicates `null` should be filled out with the `actor` from the authenticated request. Right now, permission checks are disabled, but we need to do a sweep and enable them later.

File: lib/issuer.js [hunk 18]
```diff
@@ -188,33 +230,206 @@ api.claim = function(actor, id, callback) {
...
+        res.status(200).json({location: credential.id});
+      });
+    });
 
-bedrock.events.on('bedrock-express.configure.routes', function(app) {
-  // TODO: remove
-  app.get('/', /*ensureAuthenticated,*/ function(req, res, next) {
-    res.status(200).send('Issuer REST API');
+/*
+  // this code should not be needed because of the above app.post function
```
Comment (dlongley) [inline lib/issuer.js]: Not needed right now or ever? If never, remove.

File: lib/issuer.js (hunk 5 above)
Comment (mattcollier) [inline lib/issuer.js]: Views is now required in order to pass parameters into the front end.  Removing FIXME.

File: lib/issuer.js (hunk 12 above)
Comment (mattcollier) [inline lib/issuer.js]: I'll have to revisit this again, but I believe it's checking the content type that has been provided as well as what is expected in return.

**Pull request closed without merging**